### PR TITLE
Add numeric stable option for softmax

### PR DESCRIPTION
**Pull request by yugaoTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/12661

### Problem description
ttnn.softmax is unstable when numbers are large
tested input vectors ( all pos or all negs):
[100.0, 101.0], [100.0, 1000.0], [-100.0, -101.0], [-1000.0, -100.0], [-100, -108, -99, -100, -101, -98]

as well as large random inputs

### What's changed
add numeric_stable option for softmax, for each input x, we find the max value within a row, and x-max for each row

### Checklist
- [x] Post commit CI https://github.com/tenstorrent/tt-metal/actions/runs/11035924652

### CODE REVIEW

**Review by tt-aho — APPROVED** [review]

**Comment by dmakoviichuk-tt** [comment]
@yugaoTT  please add into the description the cases where it doesn't work for now. 
And examples where it works better than old softmax.
And that you expect it to be fixed automatically with ttnn.max fix.

File: tests/ttnn/unit_tests/operations/test_softmax.py [hunk 1]
```diff
@@ -13,6 +13,98 @@
 from models.utility_functions import torch_random
 
 
+def run_softmax_stable_with_program_cache(device, batch_size, h, w):
+    torch.manual_seed(0)
+
+    torch_input_tensor = torch_random((batch_size, h, w), -1000, 1000, dtype=torch.bfloat16)
```
Comment (rfurko-tt) [inline tests/ttnn/unit_tests/operations/test_softmax.py:19]: these tests don't make sense. 

Please add specific tests to describe what specifically was fixed with stable version. Test like [-1000, 1000] will not work anyway, both in pytorch and ttnn. 

These kind of tests are main target of optimization/improvement: 
torch_input_tensor = torch.tensor([100.0, 101.0], dtype=torch.bfloat16)
torch_input_tensor = torch.tensor([100.0, 1000.0], dtype=torch.bfloat16)
torch_input_tensor = torch.tensor([-100.0, -101.0], dtype=torch.bfloat16)
torch_input_tensor = torch.tensor([-1000.0, -100.0], dtype=torch.bfloat16)

File: ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/compute/softmax.cpp [hunk 2]
```diff
@@ -62,21 +67,53 @@ void MAIN {
...
+        // create numerically stable input, x-max(x)
+        #if NUMERIC_STABLE
+            unpack_reconfig_data_format_srca(cb_in0);
+            ACQ();
+            cb_reserve_back(cb_max, 1);
+            reduce_init_delta<false, PoolType::MAX, ReduceDim::REDUCE_ROW>();
+            for (uint32_t wt = 0; wt < Wt; wt++) {
+                cb_wait_front(cb_in0, wt+1);
+                constexpr uint32_t bcast_scaler0 = 0;
+                reduce_tile<PoolType::MAX, ReduceDim::REDUCE_ROW>(cb_in0, cb_bcast_scaler, wt, bcast_scaler0, dst0);
```
Comment (rfurko-tt) [inline ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/compute/softmax.cpp:79]: could you please explain how given the information provided to `reduce_tile` will allow to ignore values placed in padding? My understanding that once ttnn::max is fixed, "stable softmax" still won't work for large negative values.

**Review by rfurko-tt — COMMENTED** [review]
Please include more information in PR description.

File: tests/ttnn/unit_tests/operations/test_softmax.py (hunk 1 above)
Comment (yugaoTT) [reply tests/ttnn/unit_tests/operations/test_softmax.py]: I added these tests above, but I disagree about the test doesn't make sense, the stable version allow large values ranges pass, while previously not passing.

Comment (rfurko-tt) [reply tests/ttnn/unit_tests/operations/test_softmax.py]: My point about test that it didn't test what was intended to. 
To provide a better perspective here:
What's the probability of all random numbers be -X (where X is large enough that previously it didn't work, but now it works). Let's assume X = 100. So the probability to generate proper test ((1000 - 100) / 2000) ** (batch_size * h * w).  

min batch = 1
min h = 32
min w = 1024
```
@pytest.mark.parametrize("batch_size", [1, 8])
@pytest.mark.parametrize("h", [32, 128])
@pytest.mark.parametrize("w", [1024, 1500])
```


P(of proper test) = 0.45 ** (32 * 1024) < 0.5 ^ 32768.
To put it in perspective, the chance is still less than flipping a coin heads 32768 times in a row.

File: ttnn/cpp/ttnn/operations/normalization/softmax/softmax_pybind.cpp [hunk 3]
```diff
@@ -150,13 +154,15 @@ void bind_normalization_softmax_in_place_operation(py::module& module) {
...
+                const std::optional<const DeviceComputeKernelConfig>& compute_kernel_config,
+                const bool numeric_stable) {
+                    return self(input_tensor, program_config, compute_kernel_config, numeric_stable);
                 },
                 py::arg("input_tensor").noconvert(),
                 py::kw_only(),
                 py::arg("program_config") = SoftmaxDefaultProgramConfig{},
-                py::arg("compute_kernel_config") = std::nullopt});
+                py::arg("compute_kernel_config") = std::nullopt,
+                py::arg("numeric_stable") = false});
```
[the remark this answers is not in the document]
Comment (yugaoTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax/softmax_pybind.cpp:165]: it doesn't neccessarily give more precision, rather depends on your input, it is input value dependant. For example if the input is already small then mins a negative max number doesn't make it better. 
It only purporse is for numerical stability. that's why I want to call it numeric_stable

I'm not sure adding even more functions to it is a better way, as we already have too many functions for softmax, and in the future we should make them into one api

File: ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/compute/softmax.cpp [hunk 4]
```diff
@@ -25,6 +26,46 @@ ALWI void REL() { release_dst(tt::DstMode::Half); }
 // The buffer for the att mask is currently sized as (1t,Wt) so we only reuse it for one HtWt-sized batch of x
 // then read another Wt tiles of mask for the next batch
 
+void calc_numeric_stable(uint32_t Wt, uint32_t ndst, uint32_t cb_in, uint32_t cb_bcast_scaler, uint32_t cb_max, uint32_t cb_out) {
```
[the remark this answers is not in the document]
Comment (yugaoTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/compute/softmax.cpp:29]: yeah I'm not sure, calc_x_minux_max ?

**Review by dmakoviichuk-tt — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/compute/softmax.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmakoviichuk-tt) [reply ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/compute/softmax.cpp:29]: I think 'stable' is a default name for this op.
